internal/controllers/core/dockercomposelogstream/reconciler.go

```diff
@@ -2,48 +2,251 @@ package dockercomposelogstream
 
 import (
 	"context"
+	"io"
+	"sync"
+	"time"
 
 	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/types"
 	ctrl "sigs.k8s.io/controller-runtime"
 	"sigs.k8s.io/controller-runtime/pkg/builder"
 	ctrlclient "sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
 	"sigs.k8s.io/controller-runtime/pkg/reconcile"
 
+	"github.com/tilt-dev/tilt/internal/controllers/apicmp"
+	"github.com/tilt-dev/tilt/internal/controllers/indexer"
+	"github.com/tilt-dev/tilt/internal/docker"
+	"github.com/tilt-dev/tilt/internal/dockercompose"
+	"github.com/tilt-dev/tilt/internal/engine/runtimelog"
 	"github.com/tilt-dev/tilt/internal/store"
+	"github.com/tilt-dev/tilt/internal/store/dockercomposeservices"
 	"github.com/tilt-dev/tilt/pkg/apis/core/v1alpha1"
+	"github.com/tilt-dev/tilt/pkg/logger"
+	"github.com/tilt-dev/tilt/pkg/model"
 )
 
 type Reconciler struct {
-	client ctrlclient.Client
-	store  store.RStore
+	client   ctrlclient.Client
+	store    store.RStore
+	dcc      dockercompose.DockerComposeClient
+	dc       docker.Client
+	requeuer *indexer.Requeuer
+
+	mu sync.Mutex
+
+	// Protected by the mutex.
+	results        map[types.NamespacedName]*Result
+	containers     map[serviceKey]*v1alpha1.DockerContainerState
+	projectWatches map[string]*ProjectWatch
 }
 
 var _ reconcile.Reconciler = &Reconciler{}
 
-func NewReconciler(client ctrlclient.Client, store store.RStore) *Reconciler {
+func NewReconciler(client ctrlclient.Client, store store.RStore,
+	dcc dockercompose.DockerComposeClient, dc docker.Client) *Reconciler {
 	return &Reconciler{
-		client: client,
-		store:  store,
+		client:         client,
+		store:          store,
+		dcc:            dcc,
+		dc:             dc.ForOrchestrator(model.OrchestratorDC),
+		projectWatches: make(map[string]*ProjectWatch),
+		results:        make(map[types.NamespacedName]*Result),
+		containers:     make(map[serviceKey]*v1alpha1.DockerContainerState),
+		requeuer:       indexer.NewRequeuer(),
 	}
 }
 
 func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
+	nn := req.NamespacedName
 	obj := &v1alpha1.DockerComposeLogStream{}
-	err := r.client.Get(ctx, req.NamespacedName, obj)
+	err := r.client.Get(ctx, nn, obj)
 	if err != nil && !apierrors.IsNotFound(err) {
 		return ctrl.Result{}, err
 	}
 
 	if apierrors.IsNotFound(err) || obj.ObjectMeta.DeletionTimestamp != nil {
+		r.clearResult(nn)
+		r.manageOwnedProjectWatches(ctx)
 		return ctrl.Result{}, nil
 	}
 
+	ctx = store.MustObjectLogHandler(ctx, r.store, obj)
+	r.manageOwnedProjectWatches(ctx)
```

Comment:
I think this is going to result in potentially arbitrary assignment of project watch logs to API objects since the context logger is for a specific obj but `manageOwnedProjectWatches` will operate across everything

Reply:
good catch! it took me a while to respond to this comment, but i got it!!!